logger_test.go

```diff
@@ -103,10 +106,132 @@ func TestLogger(t *testing.T) {
 	})
 }
 
+func TestLogger_JSON(t *testing.T) {
+	t.Run("json formatting", func(t *testing.T) {
+		var buf bytes.Buffer
+
+		logger := New(&LoggerOptions{
+			Name:       "test",
+			Output:     &buf,
+			JSONFormat: true,
+		})
+
+		logger.Info("this is test", "who", "programmer", "why", "testing is fun")
+
+		b := buf.Bytes()
+
+		var raw map[string]interface{}
+		if err := json.Unmarshal(b, &raw); err != nil {
+			t.Fatal(err)
+		}
+
+		assert.Equal(t, raw["@message"], "this is test")
+		assert.Equal(t, raw["who"], "programmer")
+		assert.Equal(t, raw["why"], "testing is fun")
+	})
+	t.Run("json formatting error type", func(t *testing.T) {
+		var buf bytes.Buffer
+
+		logger := New(&LoggerOptions{
+			Name:       "test",
+			Output:     &buf,
+			JSONFormat: true,
+		})
+
+		errMsg := errors.New("this is an error")
+		logger.Info("this is test", "who", "programmer", "err", errMsg)
+
+		b := buf.Bytes()
+
+		var raw map[string]interface{}
+		if err := json.Unmarshal(b, &raw); err != nil {
+			t.Fatal(err)
+		}
+
+		assert.Equal(t, raw["@message"], "this is test")
+		assert.Equal(t, raw["who"], "programmer")
+		assert.Equal(t, raw["err"], errMsg.Error())
+	})
+	t.Run("json formatting custom error type json marshaler", func(t *testing.T) {
+		var buf bytes.Buffer
+
+		logger := New(&LoggerOptions{
+			Name:       "test",
+			Output:     &buf,
+			JSONFormat: true,
+		})
+
+		errMsg := &customErrJSON{"this is an error"}
+		logger.Info("this is test", "who", "programmer", "err", errMsg)
+
+		b := buf.Bytes()
+
+		var raw map[string]interface{}
+		if err := json.Unmarshal(b, &raw); err != nil {
+			t.Fatal(err)
+		}
+
+		assert.Equal(t, raw["@message"], "this is test")
+		assert.Equal(t, raw["who"], "programmer")
+		assert.Equal(t, raw["err"], errMsg.Error())
+	})
+	t.Run("json formatting custom error type text marshaler", func(t *testing.T) {
+		var buf bytes.Buffer
+
+		logger := New(&LoggerOptions{
+			Name:       "test",
+			Output:     &buf,
+			JSONFormat: true,
+		})
+
+		errMsg := &customErrText{"this is an error"}
+		logger.Info("this is test", "who", "programmer", "err", errMsg)
+
+		b := buf.Bytes()
+
+		var raw map[string]interface{}
+		if err := json.Unmarshal(b, &raw); err != nil {
+			t.Fatal(err)
+		}
+
+		assert.Equal(t, raw["@message"], "this is test")
+		assert.Equal(t, raw["who"], "programmer")
+		assert.Equal(t, raw["err"], errMsg.Error())
+	})
+}
+
+type customErrJSON struct {
+	Message string
+}
+
+// error impl.
+func (c *customErrJSON) Error() string {
+	return c.Message
+}
+
+// json.Marshaler impl.
+func (c customErrJSON) MarshalJSON() ([]byte, error) {
+	return []byte(strconv.Quote(c.Message)), nil
+}
+
+type customErrText struct {
+	Message string
+}
+
+// error impl.
+func (c *customErrText) Error() string {
+	return c.Message
+}
+
+// text.Marshaler impl.
+func (c customErrText) MarshalText() ([]byte, error) {
+	return []byte(c.Message), nil
```

Comment:
If the Error function and the marshalling function return the same text it's hard to know which is actually being called in the test